src/engine/input/input.main.ts

```diff
@@ -1,91 +1,143 @@
+              console.warn("input ring buffer full");
+            }
+            break;
+          case 4:
+            if (!enqueueInputRingBuffer(inputRingBuffer, KeyCodes.Mouse3, 1)) {
+              console.warn("input ring buffer full");
+            }
+            break;
+          case 8:
+            if (!enqueueInputRingBuffer(inputRingBuffer, KeyCodes.Mouse4, 1)) {
+              console.warn("input ring buffer full");
+            }
+            break;
+          case 16:
+            if (!enqueueInputRingBuffer(inputRingBuffer, KeyCodes.Mouse5, 1)) {
+              console.warn("input ring buffer full");
+            }
+            break;
+        }
       } else {
         canvas.requestPointerLock();
       }
     }
 
     function onMouseUp({ buttons }: MouseEvent) {
       if (document.pointerLockElement === canvas) {
-        inputState.mouseButtons[0] = buttons;
+        switch (buttons) {
+          case 1:
+            if (!enqueueInputRingBuffer(inputRingBuffer, KeyCodes.Mouse1, 0)) {
+              console.warn("input ring buffer full");
+            }
+            break;
+          case 2:
+            if (!enqueueInputRingBuffer(inputRingBuffer, KeyCodes.Mouse2, 0)) {
+              console.warn("input ring buffer full");
+            }
+            break;
+          case 4:
+            if (!enqueueInputRingBuffer(inputRingBuffer, KeyCodes.Mouse3, 0)) {
+              console.warn("input ring buffer full");
+            }
+            break;
+          case 8:
+            if (!enqueueInputRingBuffer(inputRingBuffer, KeyCodes.Mouse4, 0)) {
+              console.warn("input ring buffer full");
+            }
+            break;
+          case 16:
+            if (!enqueueInputRingBuffer(inputRingBuffer, KeyCodes.Mouse5, 0)) {
+              console.warn("input ring buffer full");
+            }
+            break;
+        }
       }
     }
 
     function onKeyDown({ code }: KeyboardEvent) {
+      if (last[code]) return;
+      last[code] = true;
       if (document.pointerLockElement === canvas) {
-        flagSet(inputState.keyboard, codeToKeyCode(code), 1);
+        if (!enqueueInputRingBuffer(inputRingBuffer, codeToKeyCode(code), 1)) {
+          console.warn("input ring buffer full");
+        }
       }
     }
 
     function onKeyUp({ code }: KeyboardEvent) {
+      last[code] = false;
       if (document.pointerLockElement === canvas) {
-        flagSet(inputState.keyboard, codeToKeyCode(code), 0);
+        if (!enqueueInputRingBuffer(inputRingBuffer, codeToKeyCode(code), 0)) {
+          console.warn("input ring buffer full");
+        }
       }
     }
 
     function onMouseMove({ movementX, movementY }: MouseEvent) {
       if (document.pointerLockElement === canvas) {
-        inputState.mouseMovement[0] += movementX;
-        inputState.mouseMovement[1] += movementY;
+        if (
+          !enqueueInputRingBuffer(inputRingBuffer, KeyCodes.movementX, movementX) ||
```

Comment:
Any way to get both of these enqueued in one atomic action? Just to be sure that we don't apply `movementX` in one frame and `movementY` in the next?